File: src/models/product.ts
import { RustHelper } from '@/utils/rust.helper';
import { Accessory } from './accessory';
import { Amortization } from './amortization';
import { Bead } from './bead';
import { BeadsQuality } from './beads_quality';
import { GerdanType } from './gerdan_type';
import { Model } from './model';
import { Package } from './package';
import { Thread } from './thread';
import { normalizeNumber } from '@/utils/normalize_number';

export class Product extends Model {
  schemaPrice?: string | number | null;
  schemaAuthor?: string | null;
  name?: string | null;
  category?: string | null;
  gerdanType?: GerdanType;
  weight?: string | number | null;
  length?: string | number | null;
  width?: string | number | null;
  estimate?: string | number | null;
  surcharge?: string | number | null;
  marge?: string | number | null;
  beadsQuality?: BeadsQuality;
  beadsTotalWeight?: string | number | null;
  threadsTotalPrice?: string | number | null;
  totalPrice?: string | number | null;
  image: string | null;
  preview: string | null;
  beads: Bead[];
  accessories: Accessory[];
  threads: Thread[];
  packages: Package[];
  amortizations: Amortization[];

  constructor(product: Product) {
    super(product);
    this.schemaPrice = normalizeNumber(product.schemaPrice);
    this.name = product.name;
    this.gerdanType = product.gerdanType;
    this.weight = normalizeNumber(product.weight);
    this.length = normalizeNumber(product.length);
    this.width = normalizeNumber(product.width);
    this.estimate = normalizeNumber(product.estimate);
    this.surcharge = normalizeNumber(product.surcharge);
    this.marge = normalizeNumber(product.marge);
    this.beadsQuality = product.beadsQuality;
    this.beadsTotalWeight = normalizeNumber(product.beadsTotalWeight);
    this.threadsTotalPrice = normalizeNumber(product.threadsTotalPrice);
    this.image = product.image;
    this.preview = product.preview;
    this.totalPrice = normalizeNumber(product.totalPrice);
    this.beads = product.beads
      ? product.beads.map((item) => new Bead(item))
      : [];
    this.accessories = product.accessories
      ? product.accessories.map((item) => new Accessory(item))
      : [];
    this.threads = product.threads
      ? product.threads.map((item) => new Thread(item))
      : [];
    this.packages = product.packages
      ? product.packages.map((item) => new Package(item))
      : [];
    this.amortizations = product.amortizations
      ? product.amortizations.map((item) => new Amortization(item))
      : [];
  }

  toJS() {
    this.schemaPrice = RustHelper.toStringOptional(this.schemaPrice);
    this.name = RustHelper.toString(this.name);
    this.weight = RustHelper.toStringOptional(this.weight);
    this.length = RustHelper.toStringOptional(this.length);
    this.width = RustHelper.toStringOptional(this.width);
    this.estimate = RustHelper.toStringOptional(this.estimate);
    this.surcharge = RustHelper.toStringOptional(this.surcharge);
    this.marge = RustHelper.toStringOptional(this.marge);
    this.beadsTotalWeight = RustHelper.toStringOptional(this.beadsTotalWeight);
    this.threadsTotalPrice = RustHelper.toStringOptional(
      this.threadsTotalPrice
    );
    this.image = RustHelper.toStringOptional(this.image);
    this.preview = RustHelper.toStringOptional(this.preview);
    this.totalPrice = RustHelper.toStringOptional(this.totalPrice);
  }

  toRust() {
    this.schemaPrice = RustHelper.toNumberOptional(this.schemaPrice);
    this.name = RustHelper.toString(this.name);
    this.weight = RustHelper.toNumberOptional(this.weight);
    this.length = RustHelper.toNumberOptional(this.length);
    this.width = RustHelper.toNumberOptional(this.width);
    this.estimate = RustHelper.toNumberOptional(this.estimate);
    this.surcharge = RustHelper.toNumberOptional(this.surcharge);
    this.marge = RustHelper.toNumberOptional(this.marge);
    this.beadsTotalWeight = RustHelper.toNumberOptional(this.beadsTotalWeight);
    this.threadsTotalPrice = RustHelper.toNumberOptional(
      this.threadsTotalPrice
    );
    this.image = RustHelper.toStringOptional(this.image);
    this.preview = RustHelper.toStringOptional(this.preview);
    this.totalPrice = RustHelper.toNumberOptional(this.totalPrice);
  }
}

File: src-tauri/src/database/entities/product.rs
use rusqlite::{Error, Row};
use serde::{Deserialize, Serialize};

use crate::database::enums::{BeadsQuality, GerdanType};

use super::{
    accessory::Accessory, amortization::Amortization, bead::Bead, package::Package, thread::Thread,
};

#[derive(Serialize, Deserialize, Debug)]
#[serde(rename_all = "camelCase")]
pub struct Product {
    pub id: Option<i64>,
    pub created_at: Option<i64>,
    pub updated_at: Option<i64>,
    pub beads_quality: Option<BeadsQuality>,
    pub marge: Option<f64>,
    pub surcharge: Option<f64>,
    pub name: Option<String>,
    pub schema_price: Option<f64>,
    pub schema_author: Option<String>,
    pub length: Option<f64>,
    pub width: Option<f64>,
    pub weight: Option<f64>,
    pub estimate: Option<f64>,
    pub gerdan_type: Option<GerdanType>,
    pub beads_total_weight: Option<f64>,
    pub threads_total_price: Option<f64>,
    pub total_price: Option<f64>,
    pub image: Option<String>,
    pub preview: Option<String>,
    pub amortizations: Option<Vec<Amortization>>,
    pub beads: Option<Vec<Bead>>,
    pub accessories: Option<Vec<Accessory>>,
    pub threads: Option<Vec<Thread>>,
    pub category: Option<String>,
}

impl Product {
    pub fn from_row(row: &Row<'_>) -> Result<Product, Error> {
        Ok(Product {
            id: row.get(0)?,
            created_at: row.get(1)?,
            updated_at: row.get(2)?,
            name: row.get(3)?,
            schema_price: row.get(4)?,
            schema_author: row.get(5)?,
            length: row.get(6)?,
            width: row.get(7)?,
            weight: row.get(8)?,
            beads_quality: row.get(9)?,
            gerdan_type: row.get(10)?,
            marge: row.get(11)?,
            surcharge: row.get(12)?,
            estimate: row.get(13)?,
            beads_total_weight: row.get(14)?,
            threads_total_price: row.get(15)?,
            total_price: row.get(16)?,
            image: row.get(17)?,
            preview: row.get(18)?,
            amortizations: None,
            beads: None,
            accessories: None,
            // packages: None,
            threads: None,
            category: None,
        })
    }
}

pub struct PackageToProduct {
    pub package_id: i64,
    pub product_id: i64,
}

pub struct ThreadToProduct {
    pub thread_id: i64,
    pub product_id: i64,
}

pub struct BeadToProduct {
    pub bead_id: i64,
    pub product_id: i64,
}

pub struct AmortizationToProduct {
    pub amortization_id: i64,
    pub product_id: i64,
}

#[derive(Debug)]
pub struct AccessoryToProduct {
    pub accessory_id: i64,
    pub product_id: i64,
}
